filters/include/pcl/filters/experimental/crop_box.h

```diff
@@ -0,0 +1,191 @@
+   */
+  inline void
+  setMax(const Eigen::Vector4f& max_pt)
+  {
+    max_pt_ = max_pt;
+  }
+
+  /** \brief Get the value of the maximum point of the box, as set by the user
+   * \return the value of the internal \a max_pt parameter.
+   */
+  inline Eigen::Vector4f
+  getMax() const
+  {
+    return max_pt_;
+  }
+
+  /** \brief Set a translation value for the box
+   * \param[in] translation the (tx,ty,tz) values that the box should be translated by
+   */
+  inline void
+  setTranslation(const Eigen::Vector3f& translation)
+  {
+    translation_ = translation;
+  }
+
+  /** \brief Get the value of the box translation parameter as set by the user. */
+  Eigen::Vector3f
+  getTranslation() const
+  {
+    return translation_;
+  }
+
+  /** \brief Set a rotation value for the box
+   * \param[in] rotation the (rx,ry,rz) values that the box should be rotated by
+   */
+  inline void
+  setRotation(const Eigen::Vector3f& rotation)
+  {
+    rotation_ = rotation;
+  }
+
+  /** \brief Get the value of the box rotatation parameter, as set by the user. */
+  inline Eigen::Vector3f
+  getRotation() const
+  {
+    return rotation_;
+  }
+
+  /** \brief Set a transformation that should be applied to the cloud before filtering
+   * \param[in] transform an affine transformation that needs to be applied to the
+   * cloud before filtering
+   */
+  inline void
+  setTransform(const Eigen::Affine3f& transform)
+  {
+    transform_ = transform;
+  }
+
+  /** \brief Get the value of the transformation parameter, as set by the user. */
+  inline Eigen::Affine3f
+  getTransform() const
+  {
+    return transform_;
+  }
+
+  /** \brief Sample of point indices
+   * \param[out] indices the resultant point cloud indices
+   */
+  void
+  applyFilter(Indices& indices) override
+  {
+    Eigen::Affine3f box_transform;
+    pcl::getTransformation(translation_(0),
+                           translation_(1),
+                           translation_(2),
+                           rotation_(0),
+                           rotation_(1),
+                           rotation_(2),
+                           box_transform);
+    Eigen::Affine3f pt_transform = box_transform.inverse() * transform_;
+
+    const auto lambda = [&](const PointCloud& cloud, index_t idx) {
+      const Eigen::Vector4f pt = pt_transform * cloud.at(idx).getVector4fMap();
+      return (pt.array() >= min_pt_.array()).template head<3>().all() &&
+             (pt.array() <= max_pt_.array()).template head<3>().all();
+    };
+
+    auto filter = advanced::FunctorFilter<PointT, decltype(lambda)>(
+        lambda, this->extract_removed_indices_);
+    filter.setNegative(this->getNegative());
+    filter.setKeepOrganized(this->getKeepOrganized());
+    filter.setIndices(this->getIndices());
+    filter.setInputCloud(this->getInputCloud());
+    filter.applyFilter(indices);
+    if (this->extract_removed_indices_)
+      *removed_indices_ = *filter.getRemovedIndices(); // copy
```

Comment:
This copy is quite a tragedy, else it looks to be shorter than the previous version

Reply:
as `removed_indices_` is protected, we don't have a choice except adding a new API to return non const ptr of it

Reply:
Why can't we do `remove_indices_ = filter.getRemovedIndices();`?

Reply:
Because `getRemovedIndices()` returns `IndicesConstPtr` only (and the other overload return with copying)

Reply:
An option I would see is to add a function `applyFilter(Indices& indices, Indices& removed_indices)` to the functor filter. Indices would be placed in the given `removed_indices` instead of the member `removed_indices_`. Then that function could be called like `filter.applyFilter(indices, *removed_indices)`. Just an idea, haven't thought it through completely.

Reply:
Or we could make a change all the way up in `Filter`. 3 possibilities for better integration:
* add ctor for removed_indices
* add additional output in `filter`
* add additional function with output arg as `IndicesPtr`

Reply:
Which one seems the best way forward?

Reply:
I couldn't come up with any use case after applying to changes to `Filter`, so I would prefer to limit this to functor filter only: Either:
- @mvieth suggestion
- add additional function with output arg as `IndicesPtr` in `FunctorFilter`